src/hwdb.rs

```diff
@@ -0,0 +1,97 @@
+use std::ffi::{CString, OsStr};
+use std::io::Result;
+use std::marker::PhantomData;
+use std::os::unix::ffi::OsStrExt;
+
+use libc::c_char;
+
+use ffi;
+use list::List;
+use FromRaw;
+
+/// Rust wrapper for the `udev_hwdb` struct, which provides access to `udev`'s
+/// hardware database API.
+///
+/// Like the `udev` struct, `udev_hwdb` is refcounted and automatically managed
+/// by the Rust wrapper.
+pub struct Hwdb {
+    hwdb: *mut ffi::udev_hwdb,
+}
+
+impl Clone for Hwdb {
+    fn clone(&self) -> Self {
+        unsafe { Self::from_raw(ffi::udev_hwdb_ref(self.hwdb)) }
+    }
+}
+
+impl Drop for Hwdb {
+    fn drop(&mut self) {
+        unsafe { ffi::udev_hwdb_unref(self.hwdb) };
+    }
+}
+
+as_ffi!(Hwdb, hwdb, ffi::udev_hwdb, ffi::udev_hwdb_ref);
+
+impl Hwdb {
+    /// Creates a new Hwdb context.
+    pub fn new() -> Result<Self> {
+        // NOTE: udev_hwdb_new states that its first parameter is unused.
+        // See: https://github.com/systemd/systemd/blob/227acf00/src/libudev/libudev-hwdb.c#L29-L36
+        let ptr = try_alloc!(unsafe { ffi::udev_hwdb_new(std::ptr::null_mut()) });
+        Ok(unsafe { Self::from_raw(ptr) })
+    }
+
+    /// Queries the hardware database with the given `modalias` query,
+    /// returning an iterator over each matching entry.
+    pub fn query<S: AsRef<OsStr>>(&self, modalias: S) -> List<Hwdb> {
+        // NOTE: This unwrap can fail if someone passes a string that contains an internal NUL.
+        let modalias = CString::new(modalias.as_ref().as_bytes()).unwrap();
```

Comment:
N.B.: A user could do something like `query("bad\0query")`, which would trigger the `unwrap` here.

I'm using `AsRef<OsStr>` as a strong hint that this would be a programmer error (since an `OsStr` on Linux is defined to be null-terminated), but I don't think there's an easy way to enforce this on the type level.

Some ideas:

1. Keep the current approach (simple, but hides a potential programmer error)
1. Return a `Result<List>` (more type correct, but makes the API more tedious for something that should never happen in correct `udev` use)
1. Munge the string to turn it into a valid `CString`/`CStr` by truncating after the first `NUL` (avoids any error behavior, but also modifies the user's input)

(1) is currently implemented, but I think (3) is also reasonable (since a `NUL` inside of a modalias query would be nonsense anyways). Thoughts?

Reply:
I am in favor of keeping (1), because it will provide a clear error at runtime, which will give an indication of the problem (malformed string), while (3) might result in unexpected behavior, that is much harder to debug for the user.

But I would like to ask you to change the "unwrap" to an "expect" with a proper error description.

Reply:
Will do!